markdown/docs/addon-author-guide.md

```diff
@@ -0,0 +1,321 @@
+import Component from "ember-component";
+import $ from "jquery";
+
+let componentsToNotify = [];
+let didSetupListener = false;
+
+function setupListener() {
+  didSetupListener = true;
+  $(window).on('resize', () => {
+    componentsToNotify.forEach(c => c.windowDidResize());
+  });
+}
+
+export default Component.extend({
+  didInsertElement() {
+    if (!didSetupListener) { setupListener(); }
+    componentsToNotify.push(this);
+  },
+
+  windowDidResize() {
+    // ... do some work
+  },
+
+  willDestroy() {
+    componentsToNotify.removeObject(this);
+  }
+});
+```
+
+This lazy approach moves setup to the first time the component is used,
+improving boot time for routes in the application that don't use this
+component. It also means the component is FastBoot-compatible, because
+all of the code that touches the DOM is contained in the
+`didInsertElement()` hook, which is not invoked in FastBoot.
+
+### Third-Party Dependencies
+
+Many Ember addons wrap other JavaScript libraries in a way that makes
+them easier to use from an Ember app. For example, the
+[ivy-codemirror](https://github.com/IvyApp/ivy-codemirror) addon wraps
+the [CodeMirror code editor](https://codemirror.net/) in a component
+that makes it easy to drop into an Ember app.
+
+Sometimes the library your addon uses is itself incompatible with
+Node.js. When you include it in your `ember-cli-build.js` file, you
+include code that will prevent the app from running in FastBoot.
+
+If your addon imports third-party code and you are unable to make
+changes to it to add Node compatibility, you can add a guard to your
+`ember-cli-build.js` file to only include it in the browser build.
+
+The FastBoot addon sets the `EMBER_CLI_FASTBOOT` environment variable
+when it is building the FastBoot version of the application. You can use
+this to prevent the inclusion of a library at build time:
+
+```js
+if (!process.env.EMBER_CLI_FASTBOOT) {
+  // This will only be included in the browser build
+  app.import('some/jquery.plugin.js')
+}
+```
+
+## Browser-Only or Node-Only Initializers
+
+If your addon requires browser-specific or Node-specific initialization
+code to be run, consider using the
+[fastboot-filter-initializers](https://github.com/ronco/fastboot-filter-initializers)
+Broccoli plugin. This allows you to define target-specific
+initializers just by putting them in the right directory.
+
+For example, to write an instance initializer that only runs in the
+FastBoot environment, you'd simply put it in
+`app/instance-initializers/fastboot`.
+
+## Requiring Node Modules
+
+Some addons may need functionality when running in FastBoot that is
+normally provided by the browser. Addons can load Node modules (either
+built-in or from npm) when running in FastBoot, but only if they've been
+explicitly whitelisted first.
+
+To whitelist a dependency, you'll need to edit your addon's
+`package.json`. You probably already have an `ember-addon` field; to
```

Comment:
semicolons? only on my JS tymv :P they're usually confusing for ESL speakers, consider tweaking it